### PR TITLE
HPCGAP unification: merge `hpcgap/src/plist.c` into `src/plist.c`, replace TESTING logic by object flags

### PR DESCRIPTION
**Pull request by fingolfin**

### CODE REVIEW

**Comment by fingolfin** [comment]
I just noticed that some comments in `plist.c` still talk about the old style `TESTING`, these should be updated.

Also, we could probably get rid of `IS_TESTING_PLIST` macro.

File: src/plist.c [hunk 1]
```diff
@@ -220,8 +221,7 @@ Obj TYPE_LIST_EMPTY_IMMUTABLE;
 Obj TYPE_LIST_HOM;
 
 #define IS_TESTING_PLIST(list) \
-    (FIRST_TESTING_TNUM <= TNUM_OBJ(list) \
-  && TNUM_OBJ(list) <= LAST_TESTING_TNUM)
+    (IS_BAG_REF(list) && TEST_OBJ_FLAG(list, TESTING))
```
Comment (ChrisJefferson) [inline src/plist.c]: Why do we need to test `IS_BAG_REF` here? If it's not it's not an object, but also something horrible has happened?

File: src/plist.c [hunk 2]
```diff
@@ -728,6 +753,10 @@ Obj TypePlistHom (
         type = CALL_2ARGS( TYPE_LIST_HOM,
             family, INTOBJ_INT(tnum-T_PLIST_HOM+1) );
         ASS_LIST( types, tnum-T_PLIST_HOM+1, type );
+#ifdef HPCGAP
+        // read back element before returning it, in case another thread raced us
+	type = ELM0_LIST( types, tnum-T_PLIST_HOM+1 );
```
Comment (ChrisJefferson) [inline src/plist.c]: bad indentation here

File: src/plist.c [hunk 3]
```diff
@@ -757,6 +786,10 @@ Obj TypePlistCyc (
         type = CALL_2ARGS( TYPE_LIST_HOM,
             family, INTOBJ_INT(tnum-T_PLIST_CYC+1) );
         ASS_LIST( types, tnum-T_PLIST_CYC+1, type );
+#ifdef HPCGAP
+        // read back element before returning it, in case another thread raced us
+	type = ELM0_LIST( types, tnum-T_PLIST_CYC+1 );
```
Comment (ChrisJefferson) [inline src/plist.c]: bad indentation again?

File: src/plist.c (hunk 1 above)
Comment (fingolfin) [inline src/plist.c]: I don't think we need to; but this code is copied verbatim from the HPC-GAP side, and as usual, I tried to refrain from any changes beyond that as much as possible, to reduce the risk of regressions.

File: src/plist.c (hunk 2 above)
Comment (fingolfin) [inline src/plist.c]: the indention is fine if you set the tab width to 8. But yeah, this file is horrible in that it wildly mixes tabs and spaces. But if I fix that, the real changes get drowned in tons of tab->space changes. So I'd prefer to do it in a later PR.

**Pull request merged**